### PR TITLE
Changes to switch the LibG references

### PR DESCRIPTION
**Pull request by alfredo-pozo**
Please Note:
1. Before submitting the PR, please review [How to Contribute to Dynamo](https://github.com/DynamoDS/Dynamo/blob/master/CONTRIBUTING.md)
2. Dynamo Team will meet 1x a month to review PRs found on Github (Issues will be handled separately)
3. PRs will be reviewed from oldest to newest
4. If a reviewed PR requires changes by the owner, the owner of the PR has 30 days to respond. If the PR has seen no activity by the next session, it will be either closed by the team or depending on its utility will be taken over by someone on the team
5. PRs should use either Dynamo's default PR template or [one of these other template options](https://github.com/DynamoDS/Dynamo/wiki/Choosing-a-Pull-Request-Template) in order to be considered for review.
6. PRs that do not have one of the Dynamo PR templates completely filled out with all declarations satisfied will not be reviewed by the Dynamo team.
7. PRs made to the `DynamoRevit` repo will need to be cherry-picked into all the DynamoRevit Release branches that Dynamo supports. Contributors will be responsible for cherry-picking their reviewed commits to the other branches after a `LGTM` label is added to the PR.

### Purpose

This PR has the purpose to use the new LibG nuget packages instead of the references under "extern" folder https://jira.autodesk.com/browse/DYN-1889

### Declarations

Check these if you believe they are true

- [X] The codebase is in a better state after this PR
- [X] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [X] The level of testing this PR includes is appropriate
- [ ] User facing strings, if any, are extracted into `*.resx` files
- [X] All tests pass using the self-service CI.
- [ ] Snapshot of UI changes, if any.
- [ ] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions) and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.
- [ ] This PR modifies some build requirements and the readme is updated

### Reviewers

@aparajit-pratap 

### FYIs

@QilongTang

### CODE REVIEW

File: src/DynamoCore/DynamoCore.csproj [hunk 1]
```diff
@@ -91,9 +91,8 @@ limitations under the License.
       <HintPath>..\packages\Newtonsoft.Json.8.0.3\lib\net45\Newtonsoft.Json.dll</HintPath>
       <Private>True</Private>
     </Reference>
-    <Reference Include="ProtoGeometry, Version=1.0.0.0, Culture=neutral, processorArchitecture=MSIL">
-      <SpecificVersion>False</SpecificVersion>
-      <HintPath>..\..\extern\ProtoGeometry\ProtoGeometry.dll</HintPath>
+    <Reference Include="ProtoGeometry, Version=2.7.0.9606, Culture=neutral, processorArchitecture=AMD64">
+      <HintPath>..\packages\DynamoVisualProgramming.LibG_226_0_0.2.7.0.9606\lib\net48\ProtoGeometry.dll</HintPath>
```
Comment (aparajit-pratap) [inline src/DynamoCore/DynamoCore.csproj:95]: @alfredo-pozo can we integrate the latest LibG nuget packages that would have version bumped up to 2.9 for the LibG binaries.

File: src/DynamoCore/DynamoCore.csproj [hunk 2]
```diff
@@ -380,43 +379,40 @@ limitations under the License.
...
-      <ExternProtoGeometry Include="$(SolutionDir)..\extern\ProtoGeometry\*" />
-      <ExternProtoGeometryXml Include="$(SolutionDir)..\extern\ProtoGeometry\ProtoGeometry.XML" />
-      <ExternProtoGeometryUICulture Include="$(SolutionDir)..\extern\ProtoGeometry\$(UICulture)\*" />
-      <ExternProtoGeometryLibGLocale Include="$(SolutionDir)..\extern\ProtoGeometry\libg_locale\**\*.*" />
-      <ExternLibG224 Include="$(SolutionDir)..\extern\LibG_224_0_1\*" />
-      <ExternLibG2244 Include="$(SolutionDir)..\extern\LibG_224_4_0\*" />
-      <ExternLibG225 Include="$(SolutionDir)..\extern\LibG_225_0_0\*" />
-      <ExternLibG226 Include="$(SolutionDir)..\extern\LibG_226_0_0\*" />
-      <ExternLibG Include="$(SolutionDir)..\extern\LibG_*\*" />
+      <LibGInterface Include="$(SolutionDir)\packages\DynamoVisualProgramming.LibG_226_0_0*\tools\net48\LibG.Interface.dll" />
```
Comment (aparajit-pratap) [inline src/DynamoCore/DynamoCore.csproj:385]: What does the `*` wildcard do here?

File: src/DynamoCore/DynamoCore.csproj [hunk 3]
```diff
@@ -380,43 +379,40 @@ limitations under the License.
...
     <Copy SourceFiles="$(SolutionDir)..\doc\distrib\ADPAnalyticsConsent.rtf" DestinationFolder="$(OutputPath)" />
-    <Copy SourceFiles="@(ExternProtoGeometry)" DestinationFolder="$(OutputPath)" />
-    <Copy SourceFiles="@(ExternProtoGeometryXml)" DestinationFolder="$(OutputPath)\$(UICulture)" />
-    <Copy SourceFiles="@(ExternProtoGeometryUICulture)" DestinationFolder="$(OutputPath)\$(UICulture)" />
-    <Copy SourceFiles="@(ExternProtoGeometryLibGLocale)" DestinationFolder="$(OutputPath)\libg_locale\%(RecursiveDir)" />
-    <Copy SourceFiles="@(ExternLibG224)" DestinationFolder="$(OutputPath)libg_224_0_1\" />
-    <Copy SourceFiles="@(ExternLibG2244)" DestinationFolder="$(OutputPath)libg_224_4_0\" />
-    <Copy SourceFiles="@(ExternLibG225)" DestinationFolder="$(OutputPath)libg_225_0_0\" />
-    <Copy SourceFiles="@(ExternLibG226)" DestinationFolder="$(OutputPath)libg_226_0_0\" />
-    <Copy SourceFiles="@(ExternLibG)" DestinationFolder="$(OutputPath)\%(RecursiveDir)" />
```
Comment (aparajit-pratap) [inline src/DynamoCore/DynamoCore.csproj:419]: I wonder what this `ExternLibG` was for earlier?

File: src/DynamoCore/DynamoCore.csproj [hunk 4]
```diff
@@ -431,6 +427,12 @@ limitations under the License.
     <Copy SourceFiles="@(LegacyBinariesToNodesFolder)" DestinationFolder="$(OutputPath)\nodes\" />
   </Target>
   <Target Name="AfterBuild" Condition=" '$(OS)' != 'Unix' ">
+    <ItemGroup>
+      <ASMLibG Include="$(SolutionDir)..\extern\LibG_*\*" />
```
Comment (aparajit-pratap) [inline src/DynamoCore/DynamoCore.csproj:431]: Why is this needed? I think we are getting rid of the `extern\LibG` folders, no?

File: src/DynamoCore/DynamoCore.csproj [hunk 5]
```diff
@@ -431,6 +427,12 @@ limitations under the License.
...
+      <ExternProtoGeometry Include="$(SolutionDir)..\extern\ProtoGeometry\*" />
+    </ItemGroup>
+    <Copy SourceFiles="@(ASMLibG)" DestinationFolder="$(OutputPath)\%(RecursiveDir)" />
```
Comment (aparajit-pratap) [inline src/DynamoCore/DynamoCore.csproj:434]: As commented above, I don't think this folder is required anymore as we are getting all the necessary files from the nuget package install location.

File: src/DynamoCore/DynamoCore.csproj [hunk 6]
```diff
@@ -431,6 +427,12 @@ limitations under the License.
     <Copy SourceFiles="@(LegacyBinariesToNodesFolder)" DestinationFolder="$(OutputPath)\nodes\" />
   </Target>
   <Target Name="AfterBuild" Condition=" '$(OS)' != 'Unix' ">
+    <ItemGroup>
+      <ASMLibG Include="$(SolutionDir)..\extern\LibG_*\*" />
+      <ExternProtoGeometry Include="$(SolutionDir)..\extern\ProtoGeometry\*" />
```
Comment (aparajit-pratap) [inline src/DynamoCore/DynamoCore.csproj:432]: As discussed, this is still needed for the `protogeometry.config`, customization DLL, etc., correct?

File: src/DynamoCore/DynamoCore.csproj (hunk 1 above)
Comment (alfredo-pozo) [reply src/DynamoCore/DynamoCore.csproj]: Done with last commit

File: src/DynamoCore/DynamoCore.csproj (hunk 2 above)
Comment (alfredo-pozo) [reply src/DynamoCore/DynamoCore.csproj:385]: It takes all the files under that subfolder

On the line 395 when the CI pipeline creates the folder on runtime, on this case "extern/LibG_225_0_0" it will copy all the contents of that folder to the bin/Release/LibG_225_0_0. On the future, on the switch to 226, we won't need to change that line

File: src/DynamoCore/DynamoCore.csproj (hunk 3 above)
Comment (alfredo-pozo) [reply src/DynamoCore/DynamoCore.csproj:419]: To copy all files and subfolders with the same structure, not sure if on the past the file tree of LibG had subfolders

File: src/DynamoCore/DynamoCore.csproj (hunk 4 above)
Comment (alfredo-pozo) [reply src/DynamoCore/DynamoCore.csproj:431]: Like I said before, under the "extern" folder during the CI pipeline execution, we are creating some folders to download nuget packages like TextTemplating and ASM that are not required for a success build, but they're needed to pass all the unit tests in case you don't have an ADSK product installed (e.g. Revit), so I just moved the logic that was there to the "After Build" like we agree on our previous talk.

File: src/DynamoCore/DynamoCore.csproj (hunk 5 above)
Comment (alfredo-pozo) [reply src/DynamoCore/DynamoCore.csproj:434]: Answered on previous comment: https://github.com/DynamoDS/Dynamo/pull/11049#discussion_r476934753

File: src/DynamoCore/DynamoCore.csproj (hunk 6 above)
Comment (alfredo-pozo) [reply src/DynamoCore/DynamoCore.csproj:432]: That's right, ProtoGeometry.config and the 2 XML files.

File: src/DynamoCore/DynamoCore.csproj (hunk 3 above)
Comment (aparajit-pratap) [reply src/DynamoCore/DynamoCore.csproj:419]: I would suggest removing it.

File: src/DynamoCore/DynamoCore.csproj (hunk 4 above)
Comment (aparajit-pratap) [reply src/DynamoCore/DynamoCore.csproj:431]: I don't think this will work on dev machine setup.

File: src/DynamoCore/DynamoCore.csproj [hunk 7]
```diff
@@ -380,43 +379,40 @@ limitations under the License.
...
-      <ExternProtoGeometryLibGLocale Include="$(SolutionDir)..\extern\ProtoGeometry\libg_locale\**\*.*" />
-      <ExternLibG224 Include="$(SolutionDir)..\extern\LibG_224_0_1\*" />
-      <ExternLibG2244 Include="$(SolutionDir)..\extern\LibG_224_4_0\*" />
-      <ExternLibG225 Include="$(SolutionDir)..\extern\LibG_225_0_0\*" />
-      <ExternLibG226 Include="$(SolutionDir)..\extern\LibG_226_0_0\*" />
-      <ExternLibG Include="$(SolutionDir)..\extern\LibG_*\*" />
+      <LibGInterface Include="$(SolutionDir)\packages\DynamoVisualProgramming.LibG_226_0_0*\tools\net48\LibG.Interface.dll" />
+      <LibGProtoGeometryXml Include="$(solutiondir)\packages\dynamovisualprogramming.libg_226_0_0*\lib\net48\protogeometry.xml" />
+      <LibGProtoGeometryUICulture Include="$(SolutionDir)\packages\DynamoVisualProgramming.LibG_226_0_0*\tools\net48\$(UICulture)\*" />
+      <LibGProtoGeometryLibGLocale Include="$(SolutionDir)\packages\DynamoVisualProgramming.LibG_226_0_0.2.9.0.2375\tools\net48\libg_locale\**\*.*" />
```
Comment (aparajit-pratap) [inline src/DynamoCore/DynamoCore.csproj:388]: > That it doesn't matter which version of LibG_226_0_0 are you using that is pasted on the name of the folder, it will take that file, so we don't need to update that line every time we update the LibG nuget reference

In that case, why not use the `*` here as well?

File: src/DynamoCore/DynamoCore.csproj (hunk 5 above)
Comment (aparajit-pratap) [reply src/DynamoCore/DynamoCore.csproj:434]: > I don't think this will work on dev machine setup.

Same comment here.

File: src/DynamoCore/DynamoCore.csproj (hunk 7 above)
Comment (alfredo-pozo) [reply src/DynamoCore/DynamoCore.csproj:388]: Because on line #411 we are using a %(RecursiveDir) that doesn't work well with the * on that part

File: src/DynamoCore/DynamoCore.csproj (hunk 4 above)
Comment (alfredo-pozo) [reply src/DynamoCore/DynamoCore.csproj:431]: It works fine on a dev machine 
image

File: src/DynamoCore/DynamoCore.csproj (hunk 5 above)
Comment (alfredo-pozo) [reply src/DynamoCore/DynamoCore.csproj:434]: Same answer here https://github.com/DynamoDS/Dynamo/pull/11049#discussion_r477428595

File: src/DynamoCore/DynamoCore.csproj (hunk 7 above)
Comment (aparajit-pratap) [reply src/DynamoCore/DynamoCore.csproj:388]: So it looks like we'll have to update this csproj for every single nuget package update? I think that'll be too much. Is there a way to save us from doing that?

File: src/DynamoCore/DynamoCore.csproj (hunk 4 above)
Comment (aparajit-pratap) [reply src/DynamoCore/DynamoCore.csproj:431]: Ok, did you make sure you didn't have the extern\LibG folder locally?

Comment (alfredo-pozo) [reply src/DynamoCore/DynamoCore.csproj:431]: Yes, no LibG  folder under extern folder 
image

File: src/DynamoCore/DynamoCore.csproj (hunk 7 above)
Comment (alfredo-pozo) [reply src/DynamoCore/DynamoCore.csproj:388]: Using only the csproj element as it is, no, but now I'm thinking that we can use a similar approach of what we did on LibG where we are updating nuget.exe with powershell, pretty sure that doesn't have issues with the * on that part and it can be recursive, I'll do that change.

Comment (alfredo-pozo) [reply src/DynamoCore/DynamoCore.csproj:388]: Other option could be to change all the projects to use packagareference instead of package.config, and with that, we can create a path alias (https://docs.microsoft.com/en-us/nuget/consume-packages/package-references-in-project-files#generatepathproperty)

File: src/DynamoCore/DynamoCore.csproj [hunk 8]
```diff
@@ -380,43 +379,54 @@ limitations under the License.
   <ItemGroup>
     <Resource Include="BuiltInAndOperators\BuiltIn.Migrations.xml" />
   </ItemGroup>
-  <ItemGroup />
   <Import Project="$(MSBuildToolsPath)\Microsoft.CSharp.targets" />
+  <Choose>
+    <When Condition=" '$(OS)' != 'Unix' ">
```
Comment (QilongTang) [inline src/DynamoCore/DynamoCore.csproj:384]: What is this section for?

**Comment by QilongTang** [comment]
@alfarok Does this mean after this PR, local Dynamo builds will only have libG for ASM 226?

**Comment by aparajit-pratap** [comment]
> @alfarok Does this mean after this PR, local Dynamo builds will only have libG for ASM 226?

I think you mean @alfredo-pozo :)

File: src/DynamoCore/DynamoCore.csproj (hunk 8 above)
Comment (alfredo-pozo) [reply src/DynamoCore/DynamoCore.csproj:384]: Is to decide which powershell to use, on the Linux image for the mono-build validation we have Powershell Core available, otherwise we use the windows powershell

**Comment by alfredo-pozo** [comment]
@QilongTang We are downloading all the nuget packages of LibG: 226, 225, 224.4 and 224, and the projects have reference to 226, does that answer your question?
